experimental/dds/tree2/src/shared-tree-core/branch.ts

```diff
@@ -362,7 +361,14 @@ export class SharedTreeBranch<TEditor extends ChangeFamilyEditor, TChange> exten
 		const [newHead, change, { deletedSourceCommits, targetCommits, sourceCommits }] =
 			rebaseResult;
 
-		this.undoRedoManager.updateAfterRebase(sourceCommits, undoRedoManager);
+		if (this.undoRedoManager !== undefined) {
+			// TODO: We probably can rebase a revertible branch onto a non-revertible branch.
```

Comment:
@jenn-le if/when we change the way that we rebase the URM (get rid of cloning and instead do the commit walk), I think handling these cases in a relatively expected way without much extra code will be possible.

Reply:
I think this case should be pretty simple either way, we just don't need to update a URM since we don't have one anyways.

Reply:
Rebase non-revertible onto revertible: Do not update any URM (none to update).
Rebase revertible onto non-revertible: Currently fail. Eventually, walk commits and do the right thing. We can't do nothing here, because we might have e.g. undoable commits that get rebased into different undoable commits.
Merge revertible into non-revertible: Do not update any URM (none to update).
Merge non-revertible into revertible: Currently fail. Eventually, walk new commits and track them. We _could_ do nothing here too, if we don't want the merged commits to be part of the undo stack, but I think we could safely consider them all to be "Undoable" if they came from a branch with no URM. Do you agree?

Reply:
I personally think the second and fourth cases should also not update the URM. It feels like if a commit is created on a non-revertible, they should be treated as never revertible. It feels weird to me that "moving" these commits onto new branches would change how they behave from their original branches. Especially since we're doing this for peer branches which will never have local commits so suddenly making them revertible seems strange.

Reply:
I see. That seems like a sensible policy for "merge", but isn't it a little weird for rebase? If I make commits A, B, and C, and they are all undoable, and then I rebase onto another (not-undoable) branch that has commits D, B', and E, what do I expect to happen to my B? Do I just ignore the presence of B', and even though B has been rebased and no longer exists, I let the user undo B (which is now potentially a weird/nonsense intention, since B was replaced by B')? Or do I replace B in my undo stack with B'? Or, do I simply remove B from my undo stack?

Reply:
(I'm not sure what the right thing is in these cases, which is why I'm just failing for now. We can think about the correct behavior later, or maybe not if we never even need this case.)

Reply:
Oh that's an interesting point. Yeah, I'm not sure what the right solution is. I'm also not sure I understand what kind of situations we can get ourselves into with peer branches.

Reply:
I don't either yet. For now, I think failing is a good placeholder for "we'll implement this behavior in the future" rather than picking policy and maybe changing it. We can discuss this with Yann next week too.